docs/middleware.md

```diff
@@ -0,0 +1,109 @@
+## Adding a middleware on pallet-ibc
+
+### Overview
+
+The middleware is a module that wraps another module and adds some additional functionality to it. In the context of
+IBC, it means that the middleware is a module that is executed before/after the IBC module, and can access the packet
+data, take fees, send new packets etc. The middleware should be considered not a separate module, but rather a part of
+the module that
+it's wrapping.
+
+To add a middleware for some IBC application (e.g., ics-20 - transfer app), one needs to create a new struct that
+implements the `ibc::Module` trait and include it in the chain of other modules in runtime. Another important
+step is to make sure that the middleware [satisfies](#potential-issues) IBC packet flow to avoid unexpected behavior
+of your app (for example, a possibility for double-spending).
+
+### Implementing `Module` trait and adding to runtime
+
+As an example, we'll take a look at the existing module for taking fees on
+transfers, `ics20_fee` (`contracts/pallet-ibc/src/ics20_fee`).
+First, we need to create a struct, that contains an `inner` field, which is of type of the next module in the chain, so
+we can
+call it before/after our module execution is finished.
+
+```rust
+pub struct Ics20ServiceCharge<S: Module> {
+    inner: S,
+}
+```
+
+The `Module` contains various amount of callback methods that we can use, but probably the most important ones are
+`on_recv_packet` and `on_acknowledgement_packet`. In our case, we want to take fee from the user when the packet arrives
+and the transfer already happened, so the implementation will look something like this:
+
+```rust
+impl<S: Module> Module for Ics20ServiceCharge<S> {
+    fn on_recv_packet(
+        &self,
+        _ctx: &dyn ModuleCallbackContext,
+        output: &mut ModuleOutputBuilder,
+        packet: &mut Packet,
+        relayer: &Signer,
+    ) -> Result<Acknowledgement, Error> {
+        let mut ctx = Context::<T>::default();
+        let ack = self.inner.on_recv_packet(&mut ctx, output, packet, relayer)?;
+        let _ = Self::process_fee(&mut ctx, packet, &ack).map_err(|e| {
```

Comment:
actually the way to do this would be

`process_fee`, if it fails, we fail
if it succeeds, calls the inner
if the inner fails, we call `rollback_process_fees`

Reply:
Can you may be added as a preferred alternative? Event if it's not merged